### PR TITLE
Airbrake improvement

### PR DESCRIPTION
**Pull request by lifenod**
resolve https://github.com/theplant/appkit/issues/35 and https://github.com/theplant/appkit/issues/34

changelog https://github.com/theplant/appkit/pull/36/files#diff-4ac32a78649ca5bdd8e0ba38b7006a1e

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,14 @@
+# PR [#36](https://github.com/theplant/appkit/pull/36)
+
+* Upgrade Airbrake package.
+* Change `airbrakeNotifier.Notify` func to async.
+* `errornotifier.NewAirbrakeNotifier` func now returns `io.Closer` to
+  wait send data to Aribrake.
+* Add a new parameter `context` to `errornotifier.Notifier.Notify` allow
+  add context when call `Notify`.
+* Add log req_id and span_context to `errornotifier.NotifyOnPanic` and
```
Comment (bodhi) [inline CHANGELOG.md]: Are some of these breaking changes? So clients will have to update? If so, please list them as breaking changes here.

File: errornotifier/airbraker.go [hunk 2]
```diff
@@ -6,22 +6,31 @@ package errornotifier
...
 
 // airbrakeNotifier is a Notifier that sends messages to Airbrake,
 // constructed via NewAirbrakeNotifier
 type airbrakeNotifier struct {
-	*gobrake.Notifier
+	notifier *gobrake.Notifier
+}
+
+func (n *airbrakeNotifier) Notify(e interface{}, req *http.Request, context map[string]interface{}) {
+	notice := n.notifier.Notice(e, req, 1)
```
Comment (bodhi) [inline errornotifier/airbraker.go]: This is how the correct stack trace is captured, right?

File: errornotifier/airbraker.go [hunk 3]
```diff
@@ -6,22 +6,31 @@ package errornotifier
...
 
+	"github.com/airbrake/gobrake"
 	"github.com/theplant/appkit/log"
-	gobrake "gopkg.in/airbrake/gobrake.v1"
 )
 
 // Notifier defines an interface for reporting error
 type Notifier interface {
-	Notify(interface{}, *http.Request) error
+	Notify(err interface{}, r *http.Request, context map[string]interface{})
```
Comment (junhuif) [inline errornotifier/airbraker.go]: One note about this way:

If we use the Notifier to notify an error inside the request handler manually, then we need to put the `RequestID` or `TraceID` in manually, then the report will contains that information. 

It's nice to have that sending airbrake inside an handler will have the `RequestID` and `TraceID` included automatically.

File: errornotifier/airbraker.go (hunk 2 above)
Comment (lifenod) [reply errornotifier/airbraker.go]: right, copy from the official package: https://github.com/airbrake/gobrake/blob/7af14bfb7d8ff856c1fd7df4e0dd83caba91d170/notifier.go#L219, and I test is ok.

File: errornotifier/airbraker.go (hunk 3 above)
Comment (junhuif) [reply errornotifier/airbraker.go]: After discussed with @lifenod for a while, in Backend, the usage of Airbrake is the http server, when http request failed, the airbrake will send out. And no other usage of Airbrake in the http handlers, so I think we even with:

> If we use the Notifier to notify an error inside the request handler manually, then we need to put the RequestID or TraceID in manually, then the report will contains that information.

The result is the same. 

Let's forget this.

File: CHANGELOG.md (hunk 1 above)
Comment (lifenod) [reply CHANGELOG.md]: Added breaking changes in https://github.com/theplant/appkit/pull/36/commits/196540fa4ecb669e767e8e42fbbc8fcc797bfa24

File: errornotifier/airbraker.go (hunk 3 above)
Comment (bodhi) [reply errornotifier/airbraker.go]: I think this is good because you can add extra data if you need to.

**Pull request merged**